The double negation image

fredrik-bakke opened the pull request:
Adds the double negation image, as well as some basic properties.

martinescardo commented on source/NotionsOfDecidability/Decidable.lagda:
```diff
@@ -296,8 +296,7 @@ all-types-are-¬¬-decidable X h = claim₂ claim₁
 ¬¬-stable-if-decidable : (X : 𝓤 ̇ ) → is-decidable X → ¬¬-stable X
-¬¬-stable-if-decidable X (inl  x) = λ _ → x
-¬¬-stable-if-decidable X (inr nx) = λ h → 𝟘-elim (h nx)
+¬¬-stable-if-decidable X = ¬¬-elim
```
Good.

martinescardo commented on source/NotionsOfDecidability/Decidable.lagda:
```diff
@@ -403,3 +402,31 @@ module propositional-truncation-of-decidable-type
+∘-decidable-map f g H G F x = cases positive-case negative-case (G x)
+ where
+  positive-case : (u : fiber g x) → is-decidable (fiber (g ∘ f) x)
+  positive-case u =
```
Avoid projections with the use of pattern matching, to try to make the code more readable. There are more places below where this should also be done.

martinescardo commented on source/TypeTopology/Density.lagda:
```diff
@@ -97,3 +100,138 @@ module _ {𝓤 𝓥} {X : 𝓤 ̇ } {Y : 𝓥 ̇ } where
+                            → {X : 𝓤 ̇ } {Y : 𝓥 ̇ } (f : X → Y)
+                            → is-dense (¬¬-corestriction f)
+¬¬-corestrictions-are-dense negations-are-props f ((y , nnq) , np) =
+  nnq λ p → np (p .pr₁ , to-Σ-＝ (p .pr₂ , negations-are-props _ nnq))
```
Use pattern matching on lambda.

martinescardo commented on source/TypeTopology/Density.lagda:
```diff
@@ -97,3 +100,138 @@ module _ {𝓤 𝓥} {X : 𝓤 ̇ } {Y : 𝓥 ̇ } where
+                                                        (λ y →
+                                                         negations-are-props)
+
+¬¬-restrictions-are-left-cancellable : ({A : 𝓤 ⊔ 𝓥 ̇ } → is-prop (¬ A))
```
Because you are using `({A : 𝓤 ⊔ 𝓥 ̇ } → is-prop (¬ A))` multiple times, I propose you give a name to this, for example `negations-are-props`. Maybe place it in the FunExt file, and include a proof that FunExt implies that.

fredrik-bakke replied:
The proof that funext implies that negations are propositions already has the name `negations-are-props` in the cited file, so the axiom should probably have a different name.
I also have a different question, I see few places where funext is assumed but only `negations-are-props` is used. Should I refactor those places to assume only `negations-are-props` also?

martinescardo replied:
Just leave the uses of funext. I don't think it is a worthwhile generalization, as I don't know any natural model where negations are props but funext fails. Also, have you thought how much function extensionality one can recover from negations being props?

fredrik-bakke replied:
It was suggested to me that Gödel's dialectica may be a natural place where negations are propositions, but function extensionality fails. I don't know much about this topic beyond that however.

fredrik-bakke replied:
I factored the results on "negations are propositions" to the bottom of `Subsingletons-FunExt`. However, note that some of the results were not about funext but propext instead.

Again, apologies if this is not proper etiquette.

martinescardo commented on source/TypeTopology/Density.lagda:
```diff
@@ -97,3 +100,138 @@ module _ {𝓤 𝓥} {X : 𝓤 ̇ } {Y : 𝓥 ̇ } where
+                              → is-¬¬-stable-map (¬¬-restriction f)
+¬¬-restrictions-are-¬¬-stable negations-are-props f y nnip = ((y , a) , refl)
+ where
+ a : y ∈¬¬-image f
```
Indentation.

martinescardo reviewed (changes requested):
Thanks for this contribution. Nice!

fredrik-bakke commented:
I moved all of the current lemmas to `TypeTopology.DenseMapsProperties` as you asked. I also renamed "decidable map" to "complemented map", which seems more in line with the library's terminology.

fredrik-bakke commented:
Let me know if further changes are needed, @martinescardo. I'm still thinking to leave Lavwere's fixed point theorem for a second pull request.

martinescardo reviewed (approved):
Thanks!
